### DISCUSSION TITLE
Proof of concept for an alternative frontend to ComfyUI - ComfyBox

### DISCUSSION THREAD

**Discussion by space-nuko** [Comfy-Org/ComfyUI#440] [Ideas]
Inspired by some other visual programming tools, I whipped together a prototype of a new UI for ComfyUI. 

https://github.com/space-nuko/ComfyBox

It's still in its early days, but this is what it currently looks like

<img width="1081" alt="2023-04-07 15_25_30-http___localhost_3000_# - Chromium" src="https://user-images.githubusercontent.com/24979496/230671531-58351b84-783f-46e8-a6b2-961a3ebf7e20.png">

The idea is that you can make the execution graph as complex as you need it, and then hide it away and have all the user interface elements for the panel ready when you want to lock them down. You'd be able to organize the components on a grid, sort them, etc. Right now there's only drag and drop but I'm liking how it's coming along so far

Of course there could be extension support and custom component types with enough effort, plus I think compatibility with base ComfyUI Python nodes can be maintained as well

So this would be everything that ComfyUI already does but with a but more convenience on top if you prefer having a UI to manipulate instead of a graph like me. Any thoughts?

**Comment by WASasquatch** [comment]
Awesome! I've mentioned this a few times. It's usually how node networks are used, implemented into a larger system.

**Comment by space-nuko** [comment]
Just an FYI, I've made quite a bit of progress since last time - the widgets have now been separated from nodes and can be used to control other widgets themselves, or create custom frontend logic. I also imported anapnoe's theme for their webui fork (thanks anapnoe!)

If you don't mind some bugginess you can test out the latest developments right now, I updated the install instructions. All you have to do is extract the archive and run the included script alongside ComfyUI (no install needed)

https://github.com/space-nuko/ComfyBox

![screenshot2](https://user-images.githubusercontent.com/24979496/236601700-19079685-f3ce-45ba-b7cc-5ba34b270e6d.png)

Here is the graph for the above workflow. As you can see I've managed to reimplement ComfyUI's seed randomization using nothing but graph nodes and a custom event hook I added. Also unlike ComfyUI (as far as I know) you can run two-step workflows by reusing a previous image output (copies it from the output to the input folder), the default graph includes an example HR Fix feature

<img width="1699" alt="2023-05-06 00_16_31-http___localhost_8000_ - Chromium" src="https://user-images.githubusercontent.com/24979496/236601800-ddb6dc4e-2b3b-4b1a-a7a1-e2ee7d1a42fc.png">

**Reply by WASasquatch** [reply]
Looks like it's time to try this out. It would be neat if it somehow just read the workflow and and replaced/edited it on changed to GUI as apposed to so many extra steps with node inputs.

**Reply by space-nuko** [reply]
I think it is a tradeoff, for the extra work it takes to setup the graph the first time you can accomplish building some fairly complicated behavior customized to your needs. For example, creating a button that swaps the values of two sliders like width/height or executes only a portion of the graph 

Notably I was able to do this by taking a different approach to serializing the graph data structure than ComfyUI's frontend (saving the data in links instead of widgets) which also meant I could use vanilla litegraph's base library of nodes that are link-based

**Reply by WASasquatch** [reply]
I myself, at least, and changing the workflow often to fit agendas, or loading a new one, of hundreds. So it immediately sounds daunting. I haven't even tried it yet cause of this realization.

**Reply by space-nuko** [reply]
What would you mean by that by chance? There is a feature that autocreates the required widgets on startup so you don't have to make them all by hand. The correct min/max/step for all sliders is autoconfigured this way too

Also this design lets you have a single slider for the seed and share the same value across everything that receives a seed input. And it also solves #561 without any backend changes. You can have the LoRA name input as text and then duplicate the combo box's output to another string input like filename prefix

**Reply by WASasquatch** [reply]
> What would you mean by that by chance? There is a feature that autocreates the required widgets on startup so you don't have to make them all by hand. The correct min/max/step for all sliders is autoconfigured this way too
> 
> Also this design lets you have a single slider for the seed and share the same value across everything that receives a seed input. And it also solves #561 without any backend changes. You can have the LoRA name input as text and then duplicate the combo box's output to another string input like filename prefix

And is that customizable? I don't want all my seeds fed to one input. I have other nodes that use seeds that should be independent from first pass / second pass, etc. That would mean I'd have to go in and replug a bunch of seed inputs?

**Reply by space-nuko** [reply]
Yeah you can have one seed or however many multiple seeds. I'm also going to implement enough so you can create a checkbox like "Link Seeds" that internally turns on/off the appropriate nodes so that all seeds are linked/unlinked. This isn't something I explicitly implement, it's something you can implement to your liking using the UI components I add

**Comment by rolandwillis** [comment]
This is an awesome project and I have gotten pretty far figuring out all the various nodes and how they work and interact. One thing I have not been able to figure out is how to append to arrays. For example, I can load a JSON file which has an array in it and put it in an array node. I can then use a slider to move the index along and display the text in the array for that index. Ideally I would like to add for example the existing prompt to that array and then persist the array back to the file. All in all though I wanted to say 'Excellent Job' and please keep developing!

**Comment by halr9000** [comment]
I love the idea as I'm a heavy mobile user and the nodes are kinda terrible on a phone lol. But I don't want to go back to the dinosaur land of gradio!

**Comment by 2blackbar** [comment]
IF it comes with presets, it might make the experience more bearable, for people who dont want to tinker with nodes
Nodes are good for unusual workflows but you have to know what can go where and for that you need prope documentation which does NOT EXIST, thats the catch, eveyrone makes their nodes and you can install with manager but what noodle needs to be connected to what  is not always obvious... like custom loaders , this is the biggest issue.

**Comment by DaveScream** [comment]
I think there should be some static items at the top. For input parameters. 

Maybe it should work as "convert parameter as input"  but "place parameter in user panel"

Also it would be cool to place on user panel some parameter, for example sampler name and on nodes allow to link all sampler selectors to this element in user panel
